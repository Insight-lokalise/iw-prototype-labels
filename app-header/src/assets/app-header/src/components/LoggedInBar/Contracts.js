import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import Dropdown from '@insight/toolkit-react/lib/Dropdown/Dropdown'

import { getContractResults, switchContract, t } from 'api'
import DropdownBody from './DropdownBody'
import DropdownButton from './DropdownButton'

export default function Contracts(props) {
  const [contractDetails, setContractDetails] = useState({ results: [], totalResults: 0 })

  useEffect(() => {
    handleSearch('')
  }, [])

  function handleSearch(searchTerm) {
    getContractResults(searchTerm).then(webGroupsResponse => {
      setContractDetails(webGroupsResponse)
    })
  }

  function handleSelect(selection) {
    switchContract(selection)
  }

  function renderButton(buttonProps) {
    return (
      <DropdownButton {...buttonProps} icon="clipboard" enabled={contractDetails && contractDetails.totalResults > 1}>
        {`${t('Contract')} â€“ ${props.contract.name}`}
      </DropdownButton>
    )
  }

  return (
    <Dropdown
      id="contractDropdown"
      closeOnDropdownClick={false}
      className="c-header-account-menu__dropdown"
      dropdownMenuClassName="c-header-account-menu__dropdown-menu"
      position="full-width"
      renderButton={renderButton}
    >
      <DropdownBody
        currentItemId={props.contract.id}
        handleSearch={handleSearch}
        handleSelect={handleSelect}
        listItems={contractDetails ? contractDetails.results : []}
        type={'contract'}
      />
    </Dropdown>
  )
}

Contracts.propTypes = {
  contract: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
}
