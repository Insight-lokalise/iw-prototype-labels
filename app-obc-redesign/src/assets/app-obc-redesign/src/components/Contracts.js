/* eslint-disable jsx-a11y/label-has-for */
import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { Button, Panel } from '@insight/toolkit-react'
import {Label} from "@insight/toolkit-react/lib/Form/Components/Decorators"

import AppContext from '../context/AppProvider'
import { clickedClass, unclickedClass, PAGES } from '../constants'
import { transformContractsArray } from '../helpers'
import { fetchContracts } from "../api/us"

class Contracts extends Component {
  constructor(props, context) {
    super(props, context)  

    const { contractList = [], partnerNumber, consortiaAgreement } = context.setupDeliveryInfo
    const contractValues = transformContractsArray(props.contractInfo.contracts, contractList)      
    const myConsortiaAgreement = consortiaAgreement || props.consortiaAgreement    

    this.state = {
      contractValues,
      consortiaContracts: props.contractInfo.consortiaContracts,
      consortiaAgreement: myConsortiaAgreement,
      count: contractList ? contractList.length : 0,
      hasError: false,
      isEditable: false,
      isLoading: false,
      soldTo: partnerNumber || props.clientInfo.soldTo,
    }
  }
  
  onConfirm = () => {    
    const { soldTo } = this.state
    const { setupDeliveryInfo: { partnerNumber }, updateClientInfo, updateDeliveryFields, webGroupId, contractInfo, updateContractInfoData } = this.context
    this.setState({ isLoading: true })
    fetchContracts(soldTo, webGroupId)
      .then(({ data: { contracts, validSoldto} }) => {
        if (validSoldto) {
          updateDeliveryFields({ contractList: null, partnerNumber: soldTo })
          updateClientInfo({ contracts })
          const newContractInfoData = {...contractInfo, contracts: contracts}
          updateContractInfoData(newContractInfoData)
          const contractValues = transformContractsArray(contracts)
          this.setState({ contractValues, count: 0 })
        } else {          
          this.setState({ soldTo: partnerNumber, hasError: true })
        }
        this.setState({ isEditable: false, isLoading: false })
      })
  }

  handleClick = ({currentTarget}) => {
    const { value } = currentTarget
    this.setState(({count, contractValues}) => ({
        count: contractValues[value] ? count-1 : count+1,
        contractValues: {...contractValues, [value]: !contractValues[value] }
      }))
  }

  handleSelect = ({currentTarget}) => {
    const { value } = currentTarget
    this.setState(() => ({
        consortiaAgreement: value
      }))
  }

  viewNextStep = (contractValues, consortiaAgreement) => {
    const values = Object.keys(contractValues).filter(contractId => contractValues[contractId])
    
    this.context.updateSelectedContracts({contracts: values, consortiaAgreement})
    this.context.onSelectedTabChange(PAGES.OCCURRENCES)
  }

  render() {
    const {count, contractValues, hasError, isEditable, isLoading, soldTo, consortiaContracts, consortiaAgreement} = this.state
    const hasConsortiaContracts = Array.isArray(consortiaContracts) && consortiaContracts.length > 0

    const renderContracts = Object.keys(contractValues).map((contract) => (
      <div key={contract} className="o-grid__item  u-1/5 u-margin-bot-tiny">
        <button
          className={(contractValues[contract]) ? clickedClass : unclickedClass}
          type="button"
          onClick={this.handleClick}
          value={contract}
        >
          {contract}
        </button>
      </div>
    ))
    
    
    return (
      <div className="o-grid c-contracts">
        <div className="o-grid__item u-1/1">
          { isEditable ? (
            <div className='c-form__element'>
              <Label id="port-label" required htmlFor="soldTo">Sold To: </Label>
              <div className="c-form__control">
                <input className="c-input" id="soldTo" type="text" onChange={e => this.setState({soldTo: e.target.value})} value={soldTo} />
              </div>
              <Button color='primary' onClick={this.onConfirm} isLoading={isLoading} className="c-confirm-button">
                Confirm
              </Button>
            </div>
          ) : (
            <Fragment>
              <span className="u-bold">Sold To: </span>
              {soldTo}
              <Button
                color='primary'
                onClick={() => this.setState({ hasError: false, isEditable: true })}
                className="c-soldto-button"
              >
                Edit
              </Button>
              { hasError && (
                <span className='c-soldto-error'>Must enter a valid soldTo.</span>
              )}
            </Fragment>
          )}
        </div>
        <div className="o-grid__item u-1/1">
          <h4 className="u-margin-bot">You can select contract(s) you want to save.  If this is not required, continue to next step.</h4>
          <div className="o-grid u-margin-bot">
            { renderContracts }   
          </div>

          <Fragment>
            <label className="c-form__label">Number of selected contracts:{count}</label>

            { hasConsortiaContracts && (
              <div className="o-grid__item u-1/1">
                Consortia Agreement:
                <select className="c-consortia-select" value={consortiaAgreement} onChange={this.handleSelect}>
                  <option key={-1} value={-1}>Select a Consortia Contract...</option>
                  {consortiaContracts.map((consortia) => (
                    <option key={consortia.id} value={consortia.id}>{consortia.name}</option>
                    ))}
                </select>
              </div>
            ) }

            <Panel className="u-text-right">
              <Panel.Body>
                <Button
                  color="primary"
                  onClick={() => this.viewNextStep(contractValues, consortiaAgreement)}
                  isDisabled={isEditable}
                >
                  Continue
                </Button>
              </Panel.Body>
            </Panel>
          </Fragment>
        </div>
      </div>
    )
  }
}

Contracts.contextType = AppContext

Contracts.defaultProps = {
  consortiaAgreement: null,
  soldTo: null,
}

Contracts.propTypes = {
  contractInfo: PropTypes.shape({
    contracts: PropTypes.arrayOf(PropTypes.string).isRequired,
    consortiaContracts: PropTypes.arrayOf(PropTypes.string).isRequired,    
  }),
  consortiaAgreement: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  soldTo: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
}



export default Contracts

