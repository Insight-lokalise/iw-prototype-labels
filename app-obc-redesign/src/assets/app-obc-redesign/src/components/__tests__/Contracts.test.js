import React from "react"
import TestRenderer from 'react-test-renderer'
import Contracts from "../Contracts"

it('renders the <Contracts /> component', () => {
  const contracts = [
        "SW-0040039434",
        "SW-0040039453",
        "SW-0040040496",
        "SW-0040062808",
        "SW-0040069482",
        "SW-0040086321"
      ]

  const testElement = TestRenderer.create(
    <Contracts
      contracts={contracts}
      consortiaContracts={[]}
      soldTo={null}
      consortiaAgreement={null}
    />,
  )
  const elementSnapshot = testElement.toJSON()
  // const testInstance = testElement.root


  expect(elementSnapshot).toMatchSnapshot()
  })