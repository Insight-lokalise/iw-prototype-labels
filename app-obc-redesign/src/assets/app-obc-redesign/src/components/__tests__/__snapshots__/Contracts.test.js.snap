// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the <Contracts /> component 1`] = `
<div
  className="o-grid c-contracts"
>
  <div
    className="o-grid__item u-1/1"
  >
    <span
      className="u-bold"
    >
      Sold To: 
    </span>
    <button
      className="c-button  c-button--primary c-soldto-button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </div>
  <div
    className="o-grid__item u-1/1"
  >
    <h4
      className="u-margin-bot"
    >
      You can select contract(s) you want to save.  If this is not required, continue to next step.
    </h4>
    <div
      className="o-grid u-margin-bot"
    >
      <div
        className="o-grid__item  u-1/5 u-margin-bot-tiny"
      >
        <button
          className="c-contracts__button--unclicked c-button c-button--secondary"
          onClick={[Function]}
          type="button"
          value="SW-0040039434"
        >
          0040039434
        </button>
      </div>
      <div
        className="o-grid__item  u-1/5 u-margin-bot-tiny"
      >
        <button
          className="c-contracts__button--unclicked c-button c-button--secondary"
          onClick={[Function]}
          type="button"
          value="SW-0040039453"
        >
          0040039453
        </button>
      </div>
      <div
        className="o-grid__item  u-1/5 u-margin-bot-tiny"
      >
        <button
          className="c-contracts__button--unclicked c-button c-button--secondary"
          onClick={[Function]}
          type="button"
          value="SW-0040040496"
        >
          0040040496
        </button>
      </div>
      <div
        className="o-grid__item  u-1/5 u-margin-bot-tiny"
      >
        <button
          className="c-contracts__button--unclicked c-button c-button--secondary"
          onClick={[Function]}
          type="button"
          value="SW-0040062808"
        >
          0040062808
        </button>
      </div>
      <div
        className="o-grid__item  u-1/5 u-margin-bot-tiny"
      >
        <button
          className="c-contracts__button--unclicked c-button c-button--secondary"
          onClick={[Function]}
          type="button"
          value="SW-0040069482"
        >
          0040069482
        </button>
      </div>
      <div
        className="o-grid__item  u-1/5 u-margin-bot-tiny"
      >
        <button
          className="c-contracts__button--unclicked c-button c-button--secondary"
          onClick={[Function]}
          type="button"
          value="SW-0040086321"
        >
          0040086321
        </button>
      </div>
    </div>
    <label
      className="c-form__label"
    >
      Number of selected contracts:
      0
    </label>
    <div
      className="c-panel u-text-right"
    >
      <div
        className="o-box  c-panel__body"
      >
        <button
          className="c-button  c-button--primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Continue
        </button>
      </div>
    </div>
  </div>
</div>
`;
